redash/tasks/queries.py

```diff
@@ -379,7 +380,64 @@ def refresh_schema(data_source_id):
     logger.info(u"task=refresh_schema state=start ds_id=%s", ds.id)
     start_time = time.time()
     try:
-        ds.get_schema(refresh=True)
+        existing_tables = set()
+        schema = ds.query_runner.get_schema(get_stats=True)
+        for table in schema:
+            table_name = table['name']
+            existing_tables.add(table_name)
+
+            # Assume that there will only exist 1 table with a given name for a given data source so we use first()
+            persisted_table = models.db.session.query(
+                TableMetadata).filter(
+                TableMetadata.table_name==table_name).filter(
+                TableMetadata.data_source_id==ds.id).first()
+
+            if persisted_table:
+                models.db.session.query(TableMetadata).filter(TableMetadata.id==persisted_table.id).update({"table_exists": True})
+            else:
+                metadata = 'metadata' in table
+                persisted_table = TableMetadata(table_name=table_name, data_source_id=ds.id, column_metadata=metadata)
+                models.db.session.add(persisted_table)
+                models.db.session.flush()
+
+            existing_columns = set()
+            for i, column in enumerate(table['columns']):
+                existing_columns.add(column)
+                column_metadata = {
+                    'table_id': persisted_table.id,
+                    'column_name': column,
+                    'column_type': None,
+                    'column_example': None,
+                    'column_exists': True
+                }
+                if 'metadata' in table:
+                    column_metadata['column_type'] = table['metadata'][i]['type']
+
+                    # Note: the query example can be quite large, so we limit its size.
+                    column_example = str(table['metadata'][i]['sample'])
+                    column_metadata['column_example'] = column_example
+                    if column_example and len(column_example) > 4000:
+                        column_metadata['column_example'] = column_example[:4000] + '...'
```

Comment:
This could lead to decoding issues if the column example contains unicode characters. It'd be better to do:
```suggestion
                        column_metadata['column_example'] = u'{}...'.format('column_example[:4000])```